### PR TITLE
CNF-13832 Metadata apiv2

### PR DESCRIPTION
**Pull request by nocturnalastro**

### CODE REVIEW

**Comment by lack** [comment]
I noticed two things generally, that we should be careful about:
- The first commit has some go module bumps plus some code changes.  Let's try to keep those in separate commits (if not separate PRs when possible)
- Because we have a go module bump, you'll need to run `go work vendor` to update the vendor/ copies of our modules, and then don't forget to `git add vendor/` to get all the changes in the same commit as the go.mod changes.

**Comment by nocturnalastro** [comment]
> I noticed two things generally, that we should be careful about:
> 
>     * The first commit has some go module bumps plus some code changes.  Let's try to keep those in separate commits (if not separate PRs when possible)
> 
>     * Because we have a go module bump, you'll need to run `go work vendor` to update the vendor/ copies of our modules, and then don't forget to `git add vendor/` to get all the changes in the same commit as the go.mod changes.

I moved them to a separate PR, I was checked the sync and vendor before I started this must have forgotten to clear them.

File: docs/reference-config-guide-v2.md [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+  - name: ExamplePart2
+    ---- here goes Part2 configuration ----
+```
+
+"components" array includes objects with the following keys:
+
+- a pair at key "name" with a value of a string identifying a component required for the part
+- one of the following:
+  - `allOf`: Will cause a validation error if not all templates are matched for required templates.
+  - `allorNoneOf`: Will cause a validation error when one or more but not all templates are matched for optional blocks of templates.
```
Comment (lack) [inline docs/reference-config-guide-v2.md:60]:
```suggestion
  - `allOrNoneOf`: Will cause a validation error when some but not all of the templates are matched. Use for sets of templates where the entire set is optional, but if at least one is present then all the others must also be present.
```

File: docs/reference-config-guide-v2.md [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
+        ---- here goes ExampleComponent1 configuration ----
+  - name: ExamplePart2
+    ---- here goes Part2 configuration ----
+```
+
+"components" array includes objects with the following keys:
+
+- a pair at key "name" with a value of a string identifying a component required for the part
+- one of the following:
+  - `allOf`: Will cause a validation error if not all templates are matched for required templates.
```
Comment (lack) [inline docs/reference-config-guide-v2.md:59]:
```suggestion
  - `allOf`: Will cause a validation error if not all the templates are matched. Use for sets of required templates.
```

File: docs/reference-config-guide-v2.md [hunk 3]
```diff
@@ -0,0 +1,329 @@
...
+  - `allorNoneOf`: Will cause a validation error when one or more but not all templates are matched for optional blocks of templates.
+  - `anyOf`: Will never cause a vaildation error, for completely optional templates
```
Comment (lack) [inline docs/reference-config-guide-v2.md:61]:
```suggestion
  - `anyOf`: Will only cause validation errors for the content of matched templates. Use for completely optional templates.
```

File: docs/reference-config-guide-v2.md [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+  - `noneOf`: Will cause a vaildation error if any of the templates are matched useful to exclude CRs
```
Comment (lack) [inline docs/reference-config-guide-v2.md:62]:
```suggestion
  - `noneOf`: Will cause a validation error if any of the templates are matched. Use for excluding CRs.
```

File: docs/reference-config-guide-v2.md [hunk 5]
```diff
@@ -0,0 +1,329 @@
...
+  - `oneOf`: Will cause a vaildation error when None or more than one templates is matched, for when you want to match exactly one template in the block
```
Comment (lack) [inline docs/reference-config-guide-v2.md:63]:
```suggestion
  - `oneOf`: Will cause a validation error when none or more than one of the templates is matched. For requiring exactly one template in the set.
```

File: docs/reference-config-guide-v2.md [hunk 6]
```diff
@@ -0,0 +1,329 @@
...
+  - `anyOneOf`: Will throw a validation error if more than one is matched, for when you optionally want only one of the templates the block
```
Comment (lack) [inline docs/reference-config-guide-v2.md:64]:
```suggestion
  - `anyOneOf`: Will throw a validation error if more than one template is matched. Use for optionally allowing only one of the templates in the set.
```

File: docs/reference-config-guide-v2.md [hunk 7]
```diff
@@ -0,0 +1,329 @@
...
+
+```yaml
+# requiredTemplates contains all needed reference CRs to comply with ExampleComponent1
```
Comment (lack) [inline docs/reference-config-guide-v2.md:67]: I think this comment is wrong now?

**Review by lack — CHANGES_REQUESTED** [review]
Doc only review at first; I think we should use `allOrNoneOf` instead of `allorNoneOf` or the camelCasePolice may start knocking down our doors.

**Review by lack — APPROVED** [review]

**Pull request merged**